tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -28,62 +34,214 @@
+
+    static Trace of(boolean isObservable, String traceId) {
+        return isObservable ? new Sampled(traceId) : new Unsampled(traceId);
+    }
+
+    private static final class Sampled extends Trace {
+
+        private final Deque<OpenSpan> stack;
+
+        private Sampled(ArrayDeque<OpenSpan> stack, String traceId) {
+            super(traceId);
+            this.stack = stack;
+        }
+
+        private Sampled(String traceId) {
+            this(new ArrayDeque<>(), traceId);
+        }
+
+        @Override
+        @SuppressWarnings("ResultOfMethodCallIgnored") // Sampled traces cannot optimize this path
+        void fastStartSpan(String operation, String parentSpanId, SpanType type) {
+            startSpan(operation, parentSpanId, type);
+        }
+
+        @Override
+        @SuppressWarnings("ResultOfMethodCallIgnored") // Sampled traces cannot optimize this path
+        void fastStartSpan(String operation, SpanType type) {
+            startSpan(operation, type);
+        }
+
+        @Override
+        void push(OpenSpan span) {
+            stack.push(span);
+        }
+
+        @Override
+        Optional<OpenSpan> top() {
+            return stack.isEmpty() ? Optional.empty() : Optional.of(stack.peekFirst());
+        }
+
+        @Override
+        Optional<OpenSpan> pop() {
+            return stack.isEmpty() ? Optional.empty() : Optional.of(stack.pop());
+        }
+
+        @Override
+        boolean isEmpty() {
+            return stack.isEmpty();
+        }
+
+        @Override
+        boolean isObservable() {
+            return true;
+        }
+
+        @Override
+        Trace deepCopy() {
+            return new Sampled(new ArrayDeque<>(stack), getTraceId());
+        }
+
+        @Override
+        public String toString() {
+            return "Trace{stack=" + stack + ", isObservable=true, traceId='" + getTraceId() + "'}";
+        }
     }
 
-    @Override
-    public String toString() {
-        return "Trace{stack=" + stack + ", isObservable=" + isObservable + ", traceId='" + traceId + "'}";
+    private static final class Unsampled extends Trace {
+        private int depth;
+
+        private Unsampled(int depth, String traceId) {
+            super(traceId);
+            this.depth = depth;
+            validateDepth();
+        }
+
+        private Unsampled(String traceId) {
+            this(0, traceId);
+        }
+
+        @Override
+        void fastStartSpan(String operation, String parentSpanId, SpanType type) {
+            fastStartSpan(operation, type);
+        }
+
+        @Override
+        void fastStartSpan(String operation, SpanType type) {
+            depth++;
+        }
+
+        @Override
+        void push(OpenSpan span) {
+            depth++;
+        }
+
+        @Override
+        Optional<OpenSpan> top() {
+            return Optional.empty();
+        }
+
+        @Override
+        Optional<OpenSpan> pop() {
+            validateDepth();
```

Comment:
Hmmm in the Sampled class, calling pop() repeatedly on an empty trace just returns Optional.empty over and over again

In this case though, it will throw.  Should we make these behave the same? or is this difference important?

Reply:
pop validates that the value is positive, this should never throw unless we write a bug. I can add a comment explaining that, or remove the validation.